03-task-manager/optionalArrayMethodsReviewExtraAssignment.js

```diff
@@ -0,0 +1,291 @@
+// Review of JavaScript iterative Array methods (`.map`, `.filter` and `.forEach`)
+
+// This is an optional extra assignment
+
+///////////////////////////// Questions ///////////////////////////////////////
+
+// First, some basic background knowledge questions that we discussed during
+// the mentor session
+
+////////////////////
+// What is an array?
+//
+// An array is a data structure which contains a sequence of potentially mixed
+// data types. Structures with multiple elements are sometimes called
+// "collections"
+
+const stuff = [1, 2, 'fish', { id: 3 }];
+
+
+//////////////////////////////
+///// What is a method? /////
+//
+// It is a function that operates on a SPECIFIC data structure. This is a
+// word from Object Oriented Programming (sometimes called OOP).
+//
+// In OOP, we consider functions and data as a hybrid "thing." We call those
+// "things" objects. In JavaScript, EVERYTHING is an object, including
+// functions! But that's kinds of besides the point. The important thing is
+// the conceptual idea: data + behavior is thought of as "one thing."
+
+const object = {
+  // in OO, we group data and behavior
+  data: 1,
+  behavior: function() {
+    // "this" is a reference to the object we are inside of right now
+    this.data++;
+    console.log('OOP demo:', this.data);
+  }
+}
+
+// Every time we call `.behavior()`, the data (number) inside `object` is
+// incremented by 1, so we print "OOP demo 1", "OOP demo 2", etc.
+object.behavior()
+object.behavior()
+object.behavior()
+object.behavior()
+object.behavior()
+
+// That's why we call them "array methods" they are methods that exist on the
+// "Array" object.
+
+/////////////////////////////////////////////////
+///// What are the common JS array methods? /////
+//
+// - Array.prototype.push (does not take a callback)
+//
+// These guys all take a callback as input, and then call the callabck for
+// each item in the array
+//
+// - Array.prototype.filter
+//   - The callback should return a boolean. If the return value is true, the
+//     element becomes a member of the new array. If the return value is false,
+//     the element is filtered (removed).
+   const integers = [1, 2, 3, 4, 5];
+   // evenNumbers will be interger % 2 for each integer
+   const evenNumbers = integers.filter((integer) => {
+     return integer % 2 === 0
+   })
+
+// - Array.prototype.map
+//   - The callback recieves each item of the array. The return value is pushed
+//     into a new array
+   const numbers = [1, 2, 3];
+   const doubles = numbers.map((i) => i * 2);
+
+// - Array.prototype.forEach
+//   - `forEach` is like a "for" loop. It calls the callback for every item in
+//     the array
+   evenNumbers.forEach((thingy) => console.log('even', thingy));
+   doubles.forEach((d) => console.log('doubled!', d));
+
+// - Array.prototype.reduce
+//   - Reduce is weird & tricky
+//   - We ain't going there right now
+
```

Comment:
Hm, is reduce really that hard?  One could have them, for example, count all the letters in all the array elements.

Reply:
OK, I added some reduce examples